File: src/data/tutorials/038-construct-binary-tree-preorder-inorder.js
/**
 * Tutorial content for #038 — Construct Binary Tree from Preorder and Inorder Traversal
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given two integer arrays — preorder (root → left → right) and inorder (left → root → right) of the same binary tree — reconstruct the tree and return its root. All values are unique.`,
    example: `preorder = [3, 9, 20, 15, 7]\ninorder  = [9, 3, 15, 20, 7]\n→ preorder[0]=3 is root.\n→ In inorder, 3 is at index 1: left subtree = [9], right subtree = [15, 20, 7].\n→ Recurse: left subtree root=9, right subtree root=20.\n✅ Answer: tree [3, 9, 20, null, null, 15, 7]`,
    keyInsight: `The first element of preorder is always the root. Find that value in inorder — everything to its left is the left subtree, everything to its right is the right subtree. Recurse on each part with the corresponding slices of both arrays.`,
  },

  approaches: {
    'Recursive with Hash Map': {
      intuition: `Build a hash map from value → inorder index so you can find the root's position in O(1) instead of O(n). Then recursively build the tree: the first preorder element is the root, its inorder index splits the left and right subtrees, and the number of left elements tells you how to slice the preorder array for each subtree.`,
      steps: [
        `Build a map: inorderMap[value] = index for every element in inorder.`,
        `Define build(preStart, preEnd, inStart, inEnd) using indices into the original arrays.`,
        `If preStart > preEnd, return null (empty subtree).`,
        `Root value = preorder[preStart]. Create the root node.`,
        `Look up rootInorderIdx = inorderMap[rootVal].`,
        `Left subtree size = rootInorderIdx - inStart.`,
        `Recurse: root.left = build(preStart+1, preStart+leftSize, inStart, rootInorderIdx-1).`,
        `Recurse: root.right = build(preStart+leftSize+1, preEnd, rootInorderIdx+1, inEnd).`,
        `Return root.`,
      ],
      example: `preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]\ninorderMap: {9:0, 3:1, 15:2, 20:3, 7:4}\n\nbuild(0,4, 0,4): root=3, rootIdx=1, leftSize=1\n  left=build(1,1, 0,0): root=9, rootIdx=0, leftSize=0\n    left=build(2,1,...) → null\n    right=build(2,1,...) → null\n    return node(9)\n  right=build(2,4, 2,4): root=20, rootIdx=3, leftSize=1\n    left=build(3,3, 2,2): root=15 → node(15)\n    right=build(4,4, 4,4): root=7  → node(7)\n    return node(20)\n  return node(3)\n✅ Answer: tree [3,9,20,null,null,15,7]`,
      keyInsight: `O(n) time (O(1) lookup per node), O(n) space for the hash map. The hash map eliminates repeated linear searches through inorder, making this the optimal approach.`,
    },

    'Recursive without Hash Map': {
      intuition: `Same idea but without the hash map. For each recursive call, do a linear scan through the current inorder slice to find the root's index. Simpler to implement from scratch, but each lookup is O(n), making the overall algorithm O(n²) in the worst case (e.g., a skewed tree).`,
      steps: [
        `Define build(preorder, inorder) where each call works on sub-arrays.`,
        `If either array is empty, return null.`,
        `Root = preorder[0]. Create root node.`,
        `Find rootIdx = index of root.val in inorder (linear scan).`,
        `Left inorder slice = inorder[0..rootIdx-1], left preorder slice = preorder[1..1+rootIdx-1].`,
        `Right inorder slice = inorder[rootIdx+1..end], right preorder slice = preorder[1+rootIdx..end].`,
        `Recurse: root.left = build(leftPre, leftIn), root.right = build(rightPre, rightIn).`,
      ],
      example: `preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]\n\nbuild([3,9,20,15,7],[9,3,15,20,7]): root=3, rootIdx=1\n  left: pre=[9], in=[9] → root=9, no children → node(9)\n  right: pre=[20,15,7], in=[15,20,7] → root=20, rootIdx=1\n    left: pre=[15], in=[15] → node(15)\n    right: pre=[7], in=[7]  → node(7)\n    return node(20)\nreturn node(3)\n✅ Answer: tree [3,9,20,null,null,15,7]`,
      keyInsight: `O(n²) time in the worst case (linear search per node), O(n) space. Easier to code for an initial solution or when n is small. Use the hash map version for large inputs or production code.`,
    },

    'Iterative with Stack': {
      intuition: `Process preorder elements one by one. Keep a stack of nodes whose left subtrees are still being built. A separate inorder pointer tells you when to stop attaching left children and switch to attaching a right child. This avoids recursion and runs in O(n) time without the hash map.`,
      steps: [
        `Create root from preorder[0] and push it onto the stack. Set inIdx = 0.`,
        `For each subsequent preorder element (i from 1 to n-1): create node = new TreeNode(preorder[i]).`,
        `If stack.top().val !== inorder[inIdx]: attach node as the LEFT child of stack.top().`,
        `Otherwise: pop nodes off the stack while stack is non-empty and stack.top().val === inorder[inIdx], advancing inIdx each time. The last popped node becomes parent; attach node as its RIGHT child.`,
        `Push node onto the stack. Repeat until all preorder elements are processed.`,
      ],
      example: `preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]\n\nCreate root=3, stack=[3], inIdx=0\ni=1 (val=9): top=3, inorder[0]=9. 3≠9 → left. 3.left=9. stack=[3,9]\ni=2 (val=20): top=9, inorder[0]=9. 9==9 → pop 9(inIdx=1), top=3, inorder[1]=3. 3==3 → pop 3(inIdx=2). Stack empty. parent=3. 3.right=20. stack=[20]\ni=3 (val=15): top=20, inorder[2]=15. 20≠15 → left. 20.left=15. stack=[20,15]\ni=4 (val=7): top=15, inorder[2]=15. 15==15 → pop 15(inIdx=3), top=20, inorder[3]=20. 20==20 → pop 20(inIdx=4). parent=20. 20.right=7. stack=[7]\n✅ Answer: tree [3,9,20,null,null,15,7]`,
      keyInsight: `O(n) time, O(n) space. The most efficient iterative approach — no hash map needed. The inorder pointer is the key: it signals the boundary between left-child attachment and right-child attachment.`,
    },

    'Slicing (Intuitive)': {
      intuition: `The most readable approach: at each step, create actual sub-arrays for the left and right recursive calls rather than tracking index ranges. The first element of the preorder slice is always the root; find it in the inorder slice to compute the split. Simple to understand but creates many intermediate arrays.`,
      steps: [
        `If preorder is empty, return null.`,
        `rootVal = preorder[0]. Create root node.`,
        `Scan inorder for rootVal to find mid (its index).`,
        `Left inorder slice = inorder[0..mid-1]; left preorder slice = preorder[1..1+mid-1].`,
        `Right inorder slice = inorder[mid+1..end]; right preorder slice = preorder[1+mid..end].`,
        `root.left = buildTreeSlicing(leftPre, leftIn); root.right = buildTreeSlicing(rightPre, rightIn). Return root.`,
      ],
      example: `preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]\n\nbuildTreeSlicing([3,9,20,15,7],[9,3,15,20,7]):\n  root=3, mid=1 (index of 3 in inorder)\n  left:  pre=[9],      in=[9]       → root=9, no children → node(9)\n  right: pre=[20,15,7], in=[15,20,7] → root=20, mid=1\n    left:  pre=[15], in=[15] → node(15)\n    right: pre=[7],  in=[7]  → node(7)\n  return node(3)\n✅ Answer: tree [3,9,20,null,null,15,7]`,
      keyInsight: `O(n²) time (linear scan + array copying per node), O(n²) space (intermediate arrays). Least efficient but easiest to read and write from scratch. Good for understanding the algorithm before optimising.`,
    },

    'Standard Solution (same as Approach 1)': {
      intuition: `The canonical LeetCode submission — the hash map recursive approach packaged as the required public method buildTree(). It is functionally identical to Approach 1. Shown here to make explicit that the "standard" answer is simply the optimal recursive-with-hashmap version.`,
      steps: [
        `Build a hash map inorderMap: value → index for all inorder elements.`,
        `Call build(preorder, 0, inorder.length - 1) with a shared preorder index starting at 0.`,
        `build(pre, lo, hi): if lo > hi return null. Root = pre[preIdx++]. Mid = inorderMap[root.val].`,
        `root.left = build(pre, lo, mid-1). root.right = build(pre, mid+1, hi). Return root.`,
      ],
      example: `preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]\ninorderMap={9:0,3:1,15:2,20:3,7:4}\n\nbuild(pre,0,4): root=pre[0]=3, mid=1. preIdx=1\n  left  build(pre,0,0): root=pre[1]=9, mid=0. preIdx=2 → node(9)\n  right build(pre,2,4): root=pre[2]=20, mid=3. preIdx=3\n    left  build(pre,2,2): root=pre[3]=15 → node(15)\n    right build(pre,4,4): root=pre[4]=7  → node(7)\n  return node(3)\n✅ Answer: tree [3,9,20,null,null,15,7]`,
      keyInsight: `O(n) time, O(n) space. This is the solution you submit to LeetCode and write in interviews. The hash map makes each root lookup O(1), yielding linear overall time.`,
    },
  },
}

File: public/cpp/038-construct-binary-tree-preorder-inorder.cpp
/*
 * Problem: Construct Binary Tree from Preorder and Inorder Traversal
 * LeetCode: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
 * 
 * Description:
 * Given two integer arrays preorder and inorder where preorder is the preorder traversal of a
 * binary tree and inorder is the inorder traversal of the same tree, construct and return the
 * binary tree.
 * 
 * Example 1:
 * Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
 *        3
 *       / \
 *      9  20
 *        /  \
 *       15   7
 * Output: [3,9,20,null,null,15,7]
 * 
 * Example 2:
 * Input: preorder = [-1], inorder = [-1]
 * Output: [-1]
 * 
 * Constraints:
 * - 1 <= preorder.length <= 3000
 * - inorder.length == preorder.length
 * - -3000 <= preorder[i], inorder[i] <= 3000
 * - preorder and inorder consist of unique values.
 * - Each value of inorder also appears in preorder.
 * - preorder is guaranteed to be the preorder traversal of the tree.
 * - inorder is guaranteed to be the inorder traversal of the tree.
 * 
 * Difficulty: Medium
 * Topics: Array, Hash Table, Divide and Conquer, Tree, Binary Tree
 */

#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
using namespace std;

// Definition for a binary tree node
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /*
     * Approach 1: Recursive with Hash Map
     * 
     * Intuition:
     * - Preorder: [root | left subtree | right subtree]
     * - Inorder: [left subtree | root | right subtree]
     * - First element in preorder is always root
     * - Find root in inorder to split left and right subtrees
     * - Recursively build left and right subtrees
     * 
     * Key Insight:
     * - Preorder tells us the root
     * - Inorder tells us which nodes are in left vs right subtree
     * - Use hash map for O(1) lookup of root position in inorder
     * 
     * Algorithm:
     * 1. Build hash map: inorder value -> index
     * 2. First element in preorder is root
     * 3. Find root in inorder to determine subtree sizes
     * 4. Recursively build left subtree with:
     *    - preorder[1 : 1+leftSize]
     *    - inorder[0 : rootIndex]
     * 5. Recursively build right subtree with:
     *    - preorder[1+leftSize : end]
     *    - inorder[rootIndex+1 : end]
     * 
     * Time Complexity: O(n) - each node processed once
     * Space Complexity: O(n) - hash map + recursion stack
     */
    unordered_map<int, int> inorderMap;
    int preorderIndex;
    
    TreeNode* buildTreeHelper(vector<int>& preorder, int left, int right) {
        if (left > right) return nullptr;
        
        // Get root value from preorder
        int rootVal = preorder[preorderIndex++];
        TreeNode* root = new TreeNode(rootVal);
        
        // Find root position in inorder
        int inorderIndex = inorderMap[rootVal];
        
        // Build left and right subtrees
        root->left = buildTreeHelper(preorder, left, inorderIndex - 1);
        root->right = buildTreeHelper(preorder, inorderIndex + 1, right);
        
        return root;
    }
    
    TreeNode* buildTree_hashmap(vector<int>& preorder, vector<int>& inorder) {
        preorderIndex = 0;
        
        // Build hash map for inorder indices
        for (int i = 0; i < inorder.size(); i++) {
            inorderMap[inorder[i]] = i;
        }
        
        return buildTreeHelper(preorder, 0, inorder.size() - 1);
    }
    
    /*
     * Approach 2: Recursive without Hash Map
     * 
     * Intuition:
     * - Same logic but find root in inorder using linear search
     * - Simpler but slower
     * 
     * Time Complexity: O(n²) - O(n) for each node to find in inorder
     * Space Complexity: O(n) - recursion stack
     */
    TreeNode* buildTreeNoMap(vector<int>& preorder, vector<int>& inorder,
                             int preStart, int preEnd, int inStart, int inEnd) {
        if (preStart > preEnd || inStart > inEnd) {
            return nullptr;
        }
        
        int rootVal = preorder[preStart];
        TreeNode* root = new TreeNode(rootVal);
        
        // Find root in inorder
        int inorderIndex = inStart;
        while (inorder[inorderIndex] != rootVal) {
            inorderIndex++;
        }
        
        int leftSize = inorderIndex - inStart;
        
        root->left = buildTreeNoMap(preorder, inorder,
                                     preStart + 1, preStart + leftSize,
                                     inStart, inorderIndex - 1);
        root->right = buildTreeNoMap(preorder, inorder,
                                      preStart + leftSize + 1, preEnd,
                                      inorderIndex + 1, inEnd);
        
        return root;
    }
    
    TreeNode* buildTree_no_map(vector<int>& preorder, vector<int>& inorder) {
        return buildTreeNoMap(preorder, inorder, 0, preorder.size() - 1,
                              0, inorder.size() - 1);
    }
    
    /*
     * Approach 3: Iterative with Stack
     * 
     * Intuition:
     * - Use stack to track path from root
     * - Process preorder elements one by one
     * - Use inorder to determine when to pop from stack
     * 
     * Algorithm:
     * 1. Create root from first preorder element
     * 2. For each remaining preorder element:
     *    - If it matches inorder[inIndex], pop stack
     *    - Otherwise, add as left child
     * 3. Handle right children when popping
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(n) - stack
     */
    TreeNode* buildTree_iterative(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty()) return nullptr;
        
        TreeNode* root = new TreeNode(preorder[0]);
        stack<TreeNode*> st;
        st.push(root);
        
        int inIndex = 0;
        
        for (int i = 1; i < preorder.size(); i++) {
            TreeNode* node = new TreeNode(preorder[i]);
            TreeNode* parent = st.top();
            
            // If current preorder doesn't match inorder, go left
            if (parent->val != inorder[inIndex]) {
                parent->left = node;
            } else {
                // Pop until we find the right parent
                while (!st.empty() && st.top()->val == inorder[inIndex]) {
                    parent = st.top();
                    st.pop();
                    inIndex++;
                }
                parent->right = node;
            }
            
            st.push(node);
        }
        
        return root;
    }
    
    /*
     * Approach 4: Recursive with Explicit Ranges
     * 
     * Intuition:
     * - Pass array slices explicitly
     * - More intuitive but creates new vectors
     * 
     * Time Complexity: O(n²) - copying vectors
     * Space Complexity: O(n²) - new vectors
     */
    TreeNode* buildTree_slicing(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty()) return nullptr;
        
        int rootVal = preorder[0];
        TreeNode* root = new TreeNode(rootVal);
        
        // Find root in inorder
        int rootIndex = 0;
        while (inorder[rootIndex] != rootVal) rootIndex++;
        
        // Create slices
        vector<int> leftPreorder(preorder.begin() + 1, 
                                 preorder.begin() + 1 + rootIndex);
        vector<int> leftInorder(inorder.begin(), 
                                inorder.begin() + rootIndex);
        vector<int> rightPreorder(preorder.begin() + 1 + rootIndex, 
                                  preorder.end());
        vector<int> rightInorder(inorder.begin() + rootIndex + 1, 
                                 inorder.end());
        
        root->left = buildTree_slicing(leftPreorder, leftInorder);
        root->right = buildTree_slicing(rightPreorder, rightInorder);
        
        return root;
    }
    
    /*
     * Approach 5: Standard Solution (Hash Map - Most Common)
     * 
     * This is the most commonly used approach in interviews
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        preorderIndex = 0;
        inorderMap.clear();
        
        for (int i = 0; i < inorder.size(); i++) {
            inorderMap[inorder[i]] = i;
        }
        
        return buildTreeHelper(preorder, 0, inorder.size() - 1);
    }
};

/*
 * Helper functions for testing
 */
void printTree(TreeNode* root, string prefix = "", bool isLeft = true) {
    if (!root) return;
    
    cout << prefix;
    cout << (isLeft ? "├──" : "└──");
    cout << root->val << endl;
    
    if (root->left || root->right) {
        if (root->left) {
            printTree(root->left, prefix + (isLeft ? "│   " : "    "), true);
        }
        if (root->right) {
            printTree(root->right, prefix + (isLeft ? "│   " : "    "), false);
        }
    }
}

void deleteTree(TreeNode* root) {
    if (!root) return;
    deleteTree(root->left);
    deleteTree(root->right);
    delete root;
}

/*
 * Test Cases
 */
void runTests() {
    Solution sol;
    
    // Test Case 1: Standard tree
    cout << "Test 1 - Input: preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]" << endl;
    vector<int> pre1 = {3, 9, 20, 15, 7};
    vector<int> in1 = {9, 3, 15, 20, 7};
    
    TreeNode* result1a = sol.buildTree_hashmap(pre1, in1);
    cout << "Hash Map approach:" << endl;
    printTree(result1a);
    
    TreeNode* result1b = sol.buildTree_no_map(pre1, in1);
    cout << "No Map approach:" << endl;
    printTree(result1b);
    
    TreeNode* result1c = sol.buildTree(pre1, in1);
    cout << "Standard approach:" << endl;
    printTree(result1c);
    cout << endl;
    
    // Test Case 2: Single node
    cout << "Test 2 - Input: preorder=[-1], inorder=[-1]" << endl;
    vector<int> pre2 = {-1};
    vector<int> in2 = {-1};
    TreeNode* result2 = sol.buildTree(pre2, in2);
    printTree(result2);
    cout << endl;
    
    // Test Case 3: Linear tree (all left)
    cout << "Test 3 - Input: preorder=[1,2,3], inorder=[3,2,1]" << endl;
    vector<int> pre3 = {1, 2, 3};
    vector<int> in3 = {3, 2, 1};
    TreeNode* result3 = sol.buildTree(pre3, in3);
    printTree(result3);
    cout << endl;
    
    // Test Case 4: Linear tree (all right)
    cout << "Test 4 - Input: preorder=[1,2,3], inorder=[1,2,3]" << endl;
    vector<int> pre4 = {1, 2, 3};
    vector<int> in4 = {1, 2, 3};
    TreeNode* result4 = sol.buildTree(pre4, in4);
    printTree(result4);
    cout << endl;
    
    // Test Case 5: Balanced tree
    cout << "Test 5 - Input: preorder=[1,2,4,5,3,6,7], inorder=[4,2,5,1,6,3,7]" << endl;
    vector<int> pre5 = {1, 2, 4, 5, 3, 6, 7};
    vector<int> in5 = {4, 2, 5, 1, 6, 3, 7};
    TreeNode* result5 = sol.buildTree(pre5, in5);
    printTree(result5);
    cout << endl;
}

int main() {
    cout << "Construct Binary Tree from Preorder and Inorder Traversal" << endl;
    cout << "==========================================================" << endl << endl;
    
    runTests();
    
    return 0;
}

/*
 * Complexity Analysis Summary:
 * 
 * Approach 1 (Hash Map):
 * - Time: O(n) - each node processed once, O(1) lookup
 * - Space: O(n) - hash map + recursion stack
 * - BEST for interviews: optimal time complexity
 * 
 * Approach 2 (No Map):
 * - Time: O(n²) - O(n) to find each node in inorder
 * - Space: O(n) - recursion stack only
 * - Simpler but slower
 * 
 * Approach 3 (Iterative):
 * - Time: O(n)
 * - Space: O(n) - stack
 * - More complex but avoids recursion
 * 
 * Approach 4 (Slicing):
 * - Time: O(n²) - copying vectors
 * - Space: O(n²) - new vectors
 * - Most intuitive but inefficient
 * 
 * Key Insights:
 * 1. Preorder gives us root first
 * 2. Inorder splits tree into left and right subtrees
 * 3. Hash map enables O(1) root lookup
 * 4. Recursion naturally handles tree structure
 * 5. Must track indices carefully
 * 
 * Traversal Properties:
 * - Preorder: Root -> Left -> Right
 * - Inorder: Left -> Root -> Right
 * - Postorder: Left -> Right -> Root
 * 
 * Why This Works:
 * Example: preorder=[3,9,20,15,7], inorder=[9,3,15,20,7]
 * 
 * 1. Root is 3 (first in preorder)
 * 2. Find 3 in inorder: [9 | 3 | 15,20,7]
 * 3. Left subtree: [9], Right subtree: [15,20,7]
 * 4. In preorder: [3 | 9 | 20,15,7]
 * 5. Recursively build left with [9] and [9]
 * 6. Recursively build right with [20,15,7] and [15,20,7]
 * 
 * Index Calculation:
 * - Root index in inorder: rootIndex
 * - Left subtree size: rootIndex - inStart
 * - Left preorder: [preStart+1, preStart+leftSize]
 * - Right preorder: [preStart+leftSize+1, preEnd]
 * - Left inorder: [inStart, rootIndex-1]
 * - Right inorder: [rootIndex+1, inEnd]
 * 
 * Common Pitfalls:
 * 1. Off-by-one errors in index calculations
 * 2. Not handling empty subtrees
 * 3. Forgetting to increment preorder index
 * 4. Incorrect subtree size calculation
 * 5. Not clearing hash map between test cases
 * 
 * Interview Tips:
 * 1. Start by explaining traversal properties
 * 2. Draw tree and show how preorder/inorder relate
 * 3. Explain why we need both traversals
 * 4. Walk through index calculations carefully
 * 5. Mention hash map optimization
 * 6. Handle edge cases: single node, linear tree
 * 7. Discuss time/space complexity
 * 
 * Why We Need Both Traversals:
 * - Preorder alone: can't determine structure
 * - Inorder alone: can't determine root
 * - Together: uniquely determine tree
 * 
 * Example Walkthrough:
 * preorder = [3,9,20,15,7]
 * inorder = [9,3,15,20,7]
 * 
 * Step 1: Root = 3
 *         3
 * 
 * Step 2: Find 3 in inorder
 *         Left: [9], Right: [15,20,7]
 * 
 * Step 3: Build left subtree
 *         3
 *        /
 *       9
 * 
 * Step 4: Build right subtree (root = 20)
 *         3
 *        / \
 *       9  20
 * 
 * Step 5: Continue recursively
 *         3
 *        / \
 *       9  20
 *         /  \
 *        15   7
 * 
 * Real-World Applications:
 * 1. Reconstructing tree from serialized data
 * 2. Database query optimization
 * 3. Compiler parse tree reconstruction
 * 4. File system recovery
 * 5. Network topology reconstruction
 * 
 * Extensions and Variations:
 * 1. Construct from inorder and postorder
 * 2. Construct from preorder and postorder (not unique)
 * 3. Construct BST from preorder only
 * 4. Verify if traversals are valid
 * 5. Count possible trees from traversals
 * 
 * Related Problems:
 * - Construct Binary Tree from Inorder and Postorder
 * - Construct Binary Search Tree from Preorder
 * - Verify Preorder Serialization of Binary Tree
 * - Serialize and Deserialize Binary Tree
 * - Recover Binary Search Tree
 */

// Made with Bob
